// awilix-socketio/lib/invokers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var awilix_1 = require("awilix");
var utils_1 = require("awilix/lib/utils");
var assert_1 = __importDefault(require("assert"));
/**
 * Creates either a function invoker or a class invoker, based on whether
 * the argument can be classified as a class or not. Uses Awilix' `isClass` utility.
 *
 * @param functionOrClass
 * The function or class to invoke.
 *
 * @param opts
 * Resolver options for the class/function.
 */
function makeInvoker(functionOrClass, opts) {
    return utils_1.isClass(functionOrClass)
        ? /*tslint:disable-next-line*/
            makeClassInvoker(functionOrClass, opts)
        : /*tslint:disable-next-line*/
            makeFunctionInvoker(functionOrClass, opts);
}
exports.makeInvoker = makeInvoker;
/**
 * Returns a function that when called with a name,
 * returns another function to be used as Koa middleware.
 * That function will run `fn` with the container cradle as the
 * only parameter, and then call the `methodToInvoke` on
 * the result.
 *
 * @param, {Function} fn
 * @return {(methodToInvoke: string) => (...args) => void}
 */
function makeFunctionInvoker(fn, opts) {
    return makeResolverInvoker(awilix_1.asFunction(fn, opts));
}
exports.makeFunctionInvoker = makeFunctionInvoker;
/**
 * Same as `makeInvoker` but for classes.
 *
 * @param  {Class} Class
 * @return {(methodToInvoke: string) => (...args) => void}
 */
function makeClassInvoker(Class, opts) {
    return makeResolverInvoker(awilix_1.asClass(Class, opts));
}
exports.makeClassInvoker = makeClassInvoker;
/**
 * Returns a function that when called with a method name,
 * returns another function to be used as Koa middleware.
 * That function will run `container.build(resolver)`, and
 * then call the method on the result, passing in the Koa context
 * and `next()`.
 *
 * @param, {Resolver} resolver
 * @return {(methodToInvoke: string) => (...args) => void}
 */
function makeResolverInvoker(resolver) {
    /**
     * 2nd step is to create a method to invoke on the result
     * of the resolver.
     *
     * @param  {MethodName} methodToInvoke
     * @return {(...args) => void}
     */
    return function makeMemberInvoker(methodToInvoke) {
        /**
         * The invoker middleware.
         *
         * @param  {Socket & { container: AwilixContainer }} this
         * @param  {...*} rest
         * @return {*}
         */
        return function memberInvoker() {
            var rest = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                rest[_i] = arguments[_i];
            }
            var container = this.container;
            var resolved = container.build(resolver);
            assert_1.default(methodToInvoke, "methodToInvoke must be a valid method type, such as string, number or symbol, but was " + String(methodToInvoke));
            return resolved[methodToInvoke].apply(resolved, [this].concat(rest));
        };
    };
}
exports.makeResolverInvoker = makeResolverInvoker;
/**
 * Receives a member invoker and turns the event handler it returns into Socket.io middleware
 * Implementation is flexible in terms of arguments other than the socket but in reality there will
 * only be one other argument (the next callback)
 *
 * @export
 * @param {((this: Socket & { container: AwilixContainer }, ...rest: any[]) => any)} memberInvoker
 * @returns {*}
 */
function adaptToMiddleware(memberInvoker) {
    return function middlewareInvoker(socket) {
        var rest = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            rest[_i - 1] = arguments[_i];
        }
        return memberInvoker.bind(socket).apply(void 0, rest);
    };
}
exports.adaptToMiddleware = adaptToMiddleware;
/**
 * Injects dependencies into the middleware factory when the middleware is invoked.
 *
 * @param factory
 */
function inject(factory) {
    var resolver = getResolver(factory);
    /**
     * The invoker middleware.
     */
    return function middlewareFactoryHandler(socket) {
        var rest = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            rest[_i - 1] = arguments[_i];
        }
        var container = socket.container;
        var resolved = container.build(resolver);
        return resolved.apply(void 0, [socket].concat(rest));
    };
}
exports.inject = inject;
/**
 * Wraps or returns a resolver.
 */
function getResolver(arg) {
    if (typeof arg === 'function') {
        /*tslint:disable-next-line*/
        return awilix_1.asFunction(arg);
    }
    return arg;
}
